src/dnaio/_core.pyx

```diff
@@ -380,12 +384,56 @@ def record_names_match(header1: str, header2: str):
     fastq-dump tool (used for converting SRA files to FASTQ) appends '.1', '.2'
     and sometimes '.3' to paired-end reads if option -I is used.
     """
-    cdef:
-        str name1, name2
-
-    name1 = header1.split()[0]
-    name2 = header2.split()[0]
-    if name1 and name2 and name1[-1] in '123' and name2[-1] in '123':
-        name1 = name1[:-1]
-        name2 = name2[:-1]
-    return name1 == name2
+    if (PyUnicode_KIND(header1) != PyUnicode_1BYTE_KIND or
+        PyUnicode_KIND(header2) != PyUnicode_1BYTE_KIND):
+        raise ValueError("Strings are not valid latin-1 encoded data.")
+    # No encoding, we can compare the unicode data directly. Provided it is
+    # in 1-byte encoding, so we can find the spaces and tabs easily.
+    cdef char * header1chars = PyUnicode_1BYTE_DATA(header1)
+    cdef char * header2chars = PyUnicode_1BYTE_DATA(header2)
+    return record_name_bytes_match(header1chars, header2chars, len(header1), len(header2))
+
+
+cdef object record_name_bytes_match(char *header1chars, char* header2chars, Py_ssize_t header1length, Py_ssize_t header2length):
+    """
+    Check whether the ascii-encoded names match.
+    """
+    # Only the first part (i.e. the name without the comment) is of interest.
+    # Find the first tab or space.
+    cdef size_t header1_ends = whitespace_at(header1chars, <size_t>header1length)
+    cdef size_t header2_ends = whitespace_at(header2chars, <size_t>header2length)
+    # Quick check if the lengths match
+    if header1_ends != header2_ends:
+        return False
+
+    # check if the names end with 1, 2 or 3. (ASCII 49, 50 , 51)
+    cdef bint name1endswithnumber = 48 < header1chars[header1_ends - 1] < 52
+    cdef bint name2endswithnumber = 48 < header2chars[header1_ends - 1] < 52
+    if name1endswithnumber and name2endswithnumber:
+        # Don't compare the read pair number
+        header1_ends -= 1
+
+    # Compare the strings up to the whitespace or up to the read pair number.
+    return memcmp(<void *>header1chars, <void *>header2chars, header1_ends) == 0
+
+
+
+cdef size_t whitespace_at(char *str_ptr, size_t length):
+    # Look for the first space or tab. Return length if not found.
+    cdef char *space_ptr
+    cdef char *tab_ptr
+    cdef size_t space_at
+    cdef size_t tab_at
+    # Look for space first as it is the most likely separator
+    space_ptr = <char *>memchr(str_ptr, 32, length)
+    if space_ptr == NULL:
+        space_at = length
+    else:
+        space_at = space_ptr - str_ptr
+    # Look if there is a tab before the space
+    tab_ptr = <char *>memchr(str_ptr, 9, space_at)
+    if tab_ptr == NULL:
+        return space_at
+    if tab_ptr < space_ptr:
+        return tab_ptr - str_ptr
+    return space_at
```

Comment:
Can you please add a test for `record_names_match` that tests whether the above works with 1) no whitespace, 2) space before tab, 3) tab before space? I think it does, but this would guard against anyone trying to "simplify" the above incorrectly.